Split frontend thing-view into view and model to fix memory leak.

sogaani opened the pull request:
Fixes #561
Fixes #566

sogaani commented:
This PR seems to fix https://github.com/mozilla-iot/gateway/issues/566 too.

mrstegeman commented:
@sogaani Any chance you could rebase this? There has been a ton of churn in this area of the code for the past couple weeks, but it should be back to mostly stable.

sogaani commented:
@mrstegeman Thank you for ping. I will rebase in this week.

sogaani commented:
I have finished rebase.

## Performance check.
I recorded memory usage while 60 sec operation among things page, things-detail page and events log page. And I called garbage collection before stop recording.

The Gateway has 9 Things.

### Result
* Master branch
    * Page objects in Heap Memory After recording.
![image](https://user-images.githubusercontent.com/20137651/42378569-adb43572-8162-11e8-83f4-5b08403a4087.png)
    * Heap Memory usage and listeners while recording.
![image](https://user-images.githubusercontent.com/20137651/42378544-944a8104-8162-11e8-800d-8aeca539c538.png)


* This pull request
    * Page objects in Heap Memory After recording.
![image](https://user-images.githubusercontent.com/20137651/42378220-7c557686-8161-11e8-994c-25179d232b20.png)
    * Heap Memory usage and listeners while recording.
![image](https://user-images.githubusercontent.com/20137651/42378257-a2985c1e-8161-11e8-9bf3-e624317c5386.png)

mrstegeman commented on static/js/models/gateway-model.js:
```diff
@@ -0,0 +1,184 @@
+    }
+    return this.addQueue(() => {
+      if (!this.thingModels.has(thingId)) {
+        throw new Error(`Thing id:${thingId} allready removed`);
```
s/allready/already/

mrstegeman commented on static/js/models/gateway-model.js:
```diff
@@ -0,0 +1,184 @@
+    }
+    return this.addQueue(() => {
+      if (!this.thingModels.has(thingId)) {
+        throw new Error(`Thing id:${thingId} allready removed`);
```
s/allready/already/

mrstegeman commented on static/js/models/model.js:
```diff
@@ -0,0 +1,76 @@
+        await handler(state);
+      } catch (e) {
+        console.error(
+          `Error occurs in handler event:${event} state:${state} ${e}`
```
s/occurs/occurred/

mrstegeman commented on static/js/models/thing-model.js:
```diff
@@ -0,0 +1,325 @@
+          resolve();
+        } else {
+          console.error(`Error removing thing ${response.statusText}`);
+          reject();
```
It would be helpful to pass an error string to `reject()`. Ditto on line 86.

mrstegeman commented on static/js/models/thing-model.js:
```diff
@@ -0,0 +1,325 @@
+          resolve();
+        } else {
+          console.error(`Error updating thing ${response.statusText}`);
+          reject();
```
It would be helpful to pass an error string to `reject()`. Ditto on line 114.

mrstegeman commented on static/js/models/thing-model.js:
```diff
@@ -0,0 +1,325 @@
+   */
+  setProperty(name, value) {
+    if (!this.propertyDescriptions.hasOwnProperty(name)) {
+      return Promise.reject();
```
It would be helpful to pass an error string to `reject()`.

mrstegeman commented on static/js/models/thing-model.js:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  /**
+   * Set value to the property.
```
s/to/of/

mrstegeman commented on static/js/models/thing-model.js:
```diff
@@ -0,0 +1,325 @@
+      }
+    }).then((json) => {
+      this.onPropertyStatus(json);
+    }).catch(function(error) {
```
nit: Use an arrow function here.

sogaani commented:
I addressed review other than `an issue when editing a device` and fix conflicts.
I cannot reproduce `an issue when editing a device`...
@mrstegeman do you have reproduce step?

An issue about floorplan fixed on https://github.com/mozilla-iot/gateway/pull/1182, because the issue does not related this pull request.
